File: mobile/ios/Packages/QervonKit/Sources/QervonNetworking/QervonAPI.swift
// =============================================================================
// File:           mobile/ios/Packages/QervonKit/Sources/QervonNetworking/QervonAPI.swift
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Typed surface over every backend endpoint the Qervon Courier app uses.
//   Thin wrappers around `HTTPClient`; no business logic lives here.
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

import Foundation
import QervonCore

public struct QervonAPI: Sendable {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - Auth

    public func requestOtp(tenantSlug: String, phone: String) async throws -> OtpRequestResult {
        try await client.send(
            .post, "/v1/auth/otp/request",
            body: OtpRequestBody(tenantSlug: tenantSlug, phone: phone),
            authenticated: false
        )
    }

    public func verifyOtp(tenantSlug: String, phone: String, code: String) async throws -> AuthTokens {
        try await client.send(
            .post, "/v1/auth/otp/verify",
            body: OtpVerifyBody(tenantSlug: tenantSlug, phone: phone, code: code),
            authenticated: false
        )
    }

    public func login(email: String, password: String, tenantSlug: String) async throws -> AuthTokens {
        try await client.send(
            .post, "/v1/auth/login",
            body: LoginBody(email: email, password: password, tenantSlug: tenantSlug),
            authenticated: false
        )
    }

    public func logout(refreshToken: String) async throws {
        try await client.sendNoContent(
            .post, "/v1/auth/logout",
            body: RefreshBody(refreshToken: refreshToken),
            authenticated: false
        )
    }

    public func linkPhone(_ phone: String) async throws -> QervonUser {
        try await client.send(.post, "/v1/auth/phone", body: LinkPhoneBody(phone: phone))
    }

    /// Creates a `role=customer` account (and, if `tenantSlug` is given, a
    /// tenant membership). Returns no tokens — the caller must follow up
    /// with `login` (or `requestOtp`/`verifyOtp` once a phone is linked).
    public func register(
        email: String,
        displayName: String,
        password: String,
        tenantSlug: String?
    ) async throws {
        try await client.sendNoContent(
            .post, "/v1/auth/register",
            body: RegisterAccountBody(
                email: email, displayName: displayName, password: password, tenantSlug: tenantSlug
            ),
            authenticated: false
        )
    }

    // MARK: - Courier self-service

    public func getOwnCourier() async throws -> Courier {
        try await client.send(.get, "/v1/courier/me")
    }

    public func setAvailability(online: Bool) async throws -> Courier {
        try await client.send(.post, "/v1/courier/me/status", body: SetAvailabilityBody(online: online))
    }

    public func updateOwnLocation(_ body: UpdateLocationBody) async throws -> Courier {
        try await client.send(.post, "/v1/courier/me/location", body: body)
    }

    public func getPendingOffer() async throws -> PendingOffer? {
        try await client.send(.get, "/v1/courier/me/offer")
    }

    public func acceptOffer(orderId: UUID) async throws -> Order {
        try await client.send(.post, "/v1/courier/orders/\(orderId.uuidString)/accept", body: EmptyBody())
    }

    public func rejectOffer(orderId: UUID) async throws {
        try await client.sendNoContent(.post, "/v1/courier/orders/\(orderId.uuidString)/reject", body: EmptyBody())
    }

    public func listCourierOrders() async throws -> [Order] {
        try await client.send(.get, "/v1/courier/orders")
    }

    public func startTransit(orderId: UUID) async throws -> Order {
        try await client.send(.post, "/v1/courier/orders/\(orderId.uuidString)/pickup", body: EmptyBody())
    }

    public func deliverOrder(orderId: UUID, _ body: DeliverOrderBody) async throws -> Order {
        try await client.send(.post, "/v1/courier/orders/\(orderId.uuidString)/deliver", body: body)
    }

    /// Uploads a real delivery-proof photo (JPEG) for `orderId` to local
    /// server-side storage and returns the URL to pass as
    /// `DeliverOrderBody.photoEvidenceUrl`. See
    /// `backend/apps/api-gateway/src/http.rs`'s `upload_delivery_photo`.
    public func uploadDeliveryPhoto(orderId: UUID, jpegData: Data) async throws -> String {
        let response: UploadedFileResponse = try await client.uploadMultipart(
            "/v1/courier/orders/\(orderId.uuidString)/photo-evidence",
            fieldName: "photo",
            filename: "proof.jpg",
            mimeType: "image/jpeg",
            fileData: jpegData
        )
        return response.url
    }

    public func getWallet() async throws -> CourierWallet {
        try await client.send(.get, "/v1/courier/me/wallet")
    }

    public func getOwnRatings() async throws -> [CustomerRating] {
        try await client.send(.get, "/v1/courier/me/ratings")
    }

    // MARK: - Customer self-service

    public func getCustomerProfile() async throws -> CustomerProfile {
        try await client.send(.get, "/v1/customer/profile")
    }

    public func addAddress(
        label: String,
        latitude: Double,
        longitude: Double,
        fullAddress: String
    ) async throws -> CustomerProfile {
        try await client.send(
            .post, "/v1/customer/profile/addresses",
            body: AddAddressBody(label: label, latitude: latitude, longitude: longitude, fullAddress: fullAddress)
        )
    }

    public func removeAddress(id: UUID) async throws -> CustomerProfile {
        try await client.send(.delete, "/v1/customer/profile/addresses/\(id.uuidString)")
    }

    /// A non-binding fare estimate for a pickup/dropoff pair. The order is
    /// always charged the server's authoritative, freshly-recomputed fare
    /// at creation time — this is purely informational for the UI.
    public func getFareQuote(pickup: GeoLocation, dropoff: GeoLocation) async throws -> FareQuote {
        let path = "/v1/customer/fare-quote?pickup_latitude=\(pickup.latitude)" +
            "&pickup_longitude=\(pickup.longitude)&dropoff_latitude=\(dropoff.latitude)" +
            "&dropoff_longitude=\(dropoff.longitude)"
        return try await client.send(.get, path)
    }

    public func createOrder(_ body: CreateCustomerOrderBody) async throws -> Order {
        try await client.send(.post, "/v1/customer/orders", body: body)
    }

    public func listCustomerOrders() async throws -> [Order] {
        try await client.send(.get, "/v1/customer/orders")
    }

    public func cancelOrder(orderId: UUID) async throws -> Order {
        try await client.send(.post, "/v1/customer/orders/\(orderId.uuidString)/cancel", body: EmptyBody())
    }

    /// `nil` when there is no assigned courier yet or the order is not in a
    /// state where an ETA is meaningful — never an error for that case.
    public func getOrderEta(orderId: UUID) async throws -> EtaInfo? {
        try await client.send(.get, "/v1/customer/orders/\(orderId.uuidString)/eta")
    }

    /// A single, one-shot snapshot of the assigned courier's last reported
    /// location for this order (polled by the UI, not push-based). The
    /// backend responds with a 422 (not `null`) when there is no assigned
    /// courier yet or no location has been reported yet — both are normal,
    /// expected polling states here, not failures to surface.
    public func getOrderTracking(orderId: UUID) async throws -> LocationSnapshot? {
        do {
            return try await client.send(.get, "/v1/orders/\(orderId.uuidString)/tracking")
        } catch APIError.server {
            return nil
        }
    }

    public func rateOrder(orderId: UUID, stars: Int, comment: String?) async throws -> CustomerRating {
        try await client.send(
            .post, "/v1/customer/orders/\(orderId.uuidString)/rating",
            body: RateOrderBody(ratingStars: stars, comment: comment)
        )
    }

    public func createSupportTicket(
        orderId: UUID?,
        subject: String,
        message: String
    ) async throws -> SupportTicket {
        try await client.send(
            .post, "/v1/customer/support-tickets",
            body: OpenSupportTicketBody(orderId: orderId, subject: subject, message: message)
        )
    }

    public func listSupportTickets() async throws -> [SupportTicket] {
        try await client.send(.get, "/v1/customer/support-tickets")
    }

    public func listNotifications() async throws -> [AppNotification] {
        try await client.send(.get, "/v1/customer/notifications")
    }

    // MARK: - Native push registration

    public func registerPushDevice(platform: PushPlatform, deviceToken: String) async throws -> DevicePushToken {
        try await client.send(
            .post, "/v1/push/devices",
            body: RegisterPushDeviceBody(platform: platform.rawValue, deviceToken: deviceToken)
        )
    }

    public func listPushDevices() async throws -> [DevicePushToken] {
        try await client.send(.get, "/v1/push/devices")
    }

    public func deletePushDevice(id: UUID) async throws {
        try await client.sendNoContent(.delete, "/v1/push/devices/\(id.uuidString)")
    }
}

File: mobile/android/core/network/src/main/kotlin/com/qervon/core/network/QervonApi.kt
// =============================================================================
// File:           mobile/android/core/network/src/main/kotlin/com/qervon/core/network/QervonApi.kt
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   High-level, typed façade over `QervonApiService` — every call either
//   returns the decoded body or throws a `QervonApiException`, so feature
//   ViewModels never touch Retrofit's `Response<T>` wrapper directly.
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

package com.qervon.core.network

import com.qervon.core.common.ApiErrorBody
import com.qervon.core.common.AuthTokenStore
import com.qervon.core.common.JsonConfig
import com.qervon.core.common.QervonApiException
import com.qervon.core.common.model.Address
import com.qervon.core.common.model.AppNotification
import com.qervon.core.common.model.AuthTokens
import com.qervon.core.common.model.Courier
import com.qervon.core.common.model.CourierWallet
import com.qervon.core.common.model.CustomerProfile
import com.qervon.core.common.model.CustomerRating
import com.qervon.core.common.model.EtaInfo
import com.qervon.core.common.model.FareQuote
import com.qervon.core.common.model.LocationSnapshot
import com.qervon.core.common.model.Order
import com.qervon.core.common.model.PendingOffer
import com.qervon.core.common.model.QervonUser
import com.qervon.core.common.model.SupportTicket
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.ResponseBody
import retrofit2.Response
import java.io.IOException

class QervonApi(
    private val service: QervonApiService,
    private val tokenStore: AuthTokenStore,
    private val json: Json = JsonConfig.shared,
) {
    // ---- Auth ----

    suspend fun register(email: String, displayName: String, password: String, tenantSlug: String?) {
        call { service.register(RegisterAccountBody(email, displayName, password, tenantSlug)) }
    }

    suspend fun login(email: String, password: String, tenantSlug: String): AuthTokens {
        val body = call { service.login(LoginBody(email, password, tenantSlug)) }
        return body.toDomain().also { tokenStore.save(it) }
    }

    suspend fun requestOtp(tenantSlug: String, phone: String): String? {
        val body = call { service.requestOtp(OtpRequestBody(tenantSlug, phone)) }
        return body.devCode
    }

    suspend fun verifyOtp(tenantSlug: String, phone: String, code: String): AuthTokens {
        val body = call { service.verifyOtp(OtpVerifyBody(tenantSlug, phone, code)) }
        return body.toDomain().also { tokenStore.save(it) }
    }

    suspend fun setPhone(phone: String): QervonUser = call { service.setPhone(SetPhoneBody(phone)) }

    fun logout() = tokenStore.clear()

    // ---- Courier ----

    suspend fun getOwnCourier(): Courier = call { service.getOwnCourier() }

    suspend fun getOwnWallet(): CourierWallet = call { service.getOwnWallet() }

    suspend fun getOwnRatings(): List<CustomerRating> = call { service.getOwnRatings() }

    suspend fun setOwnAvailability(online: Boolean): Courier =
        call { service.setOwnAvailability(SetCourierAvailabilityBody(online)) }

    suspend fun updateOwnLocation(latitude: Double, longitude: Double, speedKmh: Double?, batteryPct: Int?): Courier =
        call { service.updateOwnLocation(UpdateLocationBody(latitude, longitude, speedKmh, batteryPct)) }

    suspend fun getOwnPendingOffer(): PendingOffer? =
        fetchOptionalJson(PendingOffer.serializer()) { service.getOwnPendingOffer() }

    suspend fun listCourierOrders(): List<Order> = call { service.listCourierOrders() }

    suspend fun acceptOffer(orderId: String): Order = call { service.acceptOffer(orderId) }

    suspend fun rejectOffer(orderId: String) {
        call { service.rejectOffer(orderId) }
    }

    suspend fun pickupOrder(orderId: String, pickupPhotoEvidenceUrl: String): Order =
        call { service.pickupOrder(orderId, CompletePickupBody(pickupPhotoEvidenceUrl)) }

    suspend fun deliverOrder(
        orderId: String,
        recipientName: String,
        qrBarcodeVerified: Boolean,
        digitalSignatureBase64: String?,
        photoEvidenceUrl: String?,
        paymentCollected: Boolean,
    ): Order = call {
        service.deliverOrder(
            orderId,
            CompleteDeliveryBody(recipientName, qrBarcodeVerified, digitalSignatureBase64, photoEvidenceUrl, paymentCollected),
        )
    }

    /** Uploads a real delivery-proof photo (JPEG) for [orderId] to local
     * server-side storage and returns the URL to pass as
     * `photoEvidenceUrl` on [deliverOrder]. See
     * `backend/apps/api-gateway/src/http.rs`'s `upload_delivery_photo`. */
    suspend fun uploadOrderEvidencePhoto(orderId: String, jpegBytes: ByteArray): String {
        val body = jpegBytes.toRequestBody("image/jpeg".toMediaType())
        val part = MultipartBody.Part.createFormData("photo", "proof.jpg", body)
        return call { service.uploadDeliveryPhoto(orderId, part) }.url
    }

    suspend fun uploadDeliveryPhoto(orderId: String, jpegBytes: ByteArray): String =
        uploadOrderEvidencePhoto(orderId, jpegBytes)

    // ---- Customer ----

    suspend fun getCustomerProfile(): CustomerProfile = call { service.getCustomerProfile() }

    suspend fun addAddress(label: String, latitude: Double, longitude: Double, fullAddress: String): CustomerProfile =
        call { service.addAddress(CreateCustomerAddressBody(label, latitude, longitude, fullAddress)) }

    suspend fun removeAddress(addressId: String): CustomerProfile = call { service.removeAddress(addressId) }

    suspend fun getFareQuote(pickup: Address, dropoff: Address): FareQuote = call {
        service.getFareQuote(pickup.latitude, pickup.longitude, dropoff.latitude, dropoff.longitude)
    }

    suspend fun createOrder(
        pickup: Address,
        dropoff: Address,
        couponCode: String?,
        paymentMethod: String?,
        deliveryNote: String?,
        contactPhone: String?,
    ): Order = call {
        service.createOrder(CreateCustomerOrderBody(pickup, dropoff, couponCode, paymentMethod, deliveryNote, contactPhone))
    }

    suspend fun listCustomerOrders(): List<Order> = call { service.listCustomerOrders() }

    suspend fun cancelOrder(orderId: String): Order = call { service.cancelOrder(orderId) }

    /** Returns null when the order has no assigned courier yet or the
     * courier hasn't reported a location — an expected, non-error state
     * the backend surfaces as `Json<Option<EtaResponse>>`. */
    suspend fun getOrderEta(orderId: String): EtaInfo? =
        fetchOptionalJson(EtaInfo.serializer()) { service.getOrderEta(orderId) }

    suspend fun rateOrder(orderId: String, ratingStars: Int, comment: String?): CustomerRating =
        call { service.rateOrder(orderId, RateOrderBody(ratingStars, comment)) }

    suspend fun createSupportTicket(orderId: String?, subject: String, message: String): SupportTicket =
        call { service.createSupportTicket(OpenSupportTicketBody(orderId, subject, message)) }

    suspend fun listSupportTickets(): List<SupportTicket> = call { service.listSupportTickets() }

    suspend fun listNotifications(): List<AppNotification> = call { service.listNotifications() }

    // ---- Shared ----

    suspend fun getOrderTracking(orderId: String): LocationSnapshot = call { service.getOrderTracking(orderId) }

    // ---- Internals ----

    private fun AuthResponseBody.toDomain() = AuthTokens(accessToken, refreshToken, tokenType, expiresInSeconds.toInt())

    /** Reads a raw `Response<ResponseBody>`, treating a literal JSON
     * `null` (or an empty body) as "absent" and anything else as the
     * decoded [serializer] payload. Used for the handful of endpoints
     * that respond with `Json<Option<T>>` on the Rust side. */
    private suspend fun <T : Any> fetchOptionalJson(
        serializer: kotlinx.serialization.KSerializer<T>,
        request: suspend () -> Response<ResponseBody>,
    ): T? {
        val response = try {
            request()
        } catch (error: IOException) {
            throw QervonApiException.Transport(error)
        }
        if (!response.isSuccessful) {
            throw response.toApiException(json)
        }
        val text = response.body()?.string()?.trim()
        if (text.isNullOrEmpty() || text == "null") {
            return null
        }
        return try {
            json.decodeFromString(serializer, text)
        } catch (error: SerializationException) {
            throw QervonApiException.Decoding(error)
        }
    }

    private suspend inline fun <reified T> call(crossinline request: suspend () -> Response<T>): T {
        val response = try {
            request()
        } catch (error: IOException) {
            throw QervonApiException.Transport(error)
        }
        if (!response.isSuccessful) {
            throw response.toApiException(json)
        }
        val body = response.body()
        if (body == null) {
            @Suppress("UNCHECKED_CAST")
            return Unit as T
        }
        return body
    }
}

private fun <T> Response<T>.toApiException(json: Json): QervonApiException {
    val errorText = errorBody()?.string()
    if (errorText.isNullOrBlank()) {
        return QervonApiException.UnexpectedStatus(code())
    }
    return try {
        val decoded = json.decodeFromString(ApiErrorBody.serializer(), errorText)
        QervonApiException.Server(decoded.status, decoded.detail)
    } catch (_: SerializationException) {
        QervonApiException.UnexpectedStatus(code())
    }
}
